controllers/spannerautoscaler_controller_test.go

```diff
@@ -0,0 +1,403 @@
+)
+
+var _ = DescribeTable("Next Valid ProcessingUnits",
+	func(input, want int) {
+		got := nextValidProcessingUnits(int32(input))
+		Expect(got).To(Equal(int32(want)))
+	},
+	EntryDescription("Next valid ProcessingUnit for %d is %d"),
+	Entry(nil, 0, 100),
+	Entry(nil, 99, 100),
+	Entry(nil, 100, 200),
+	Entry(nil, 900, 1000),
+	Entry(nil, 1000, 2000),
+	Entry(nil, 1999, 2000),
+	Entry(nil, 2000, 3000),
+)
+
+var _ = Describe("Fetch Credentials", func() {
+
+	type testResult struct {
+		want        *syncer.Credentials
+		expectedErr error
+	}
+
+	var (
+		result           testResult
+		testSecret       *corev1.Secret
+		testSecretRefKey string
+		testSecretRefVal string
+		random           string
+		testResource     *spannerv1alpha1.SpannerAutoscaler
+		testReconciler   *SpannerAutoscalerReconciler
+	)
+
+	BeforeEach(func() {
+		random = uuid.NewString()
+		testSecretRefKey = "service-account"
+		testSecretRefVal = `{"foo":"bar"}`
+		result = testResult{}
+
+		testSecretName := "secret-" + random
+		testResourceName := "spanner-autoscaler-" + random
+
+		testSecret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      testSecretName,
+				Namespace: namespace,
+			},
+			StringData: map[string]string{testSecretRefKey: testSecretRefVal},
+		}
+
+		testResource = &spannerv1alpha1.SpannerAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      testResourceName,
+				Namespace: namespace,
+			},
+			Spec: spannerv1alpha1.SpannerAutoscalerSpec{
+				ServiceAccountSecretRef: &spannerv1alpha1.ServiceAccountSecretRef{
+					Namespace: pointer.String(namespace),
+					Name:      pointer.String(testSecretName),
+					Key:       pointer.String(testSecretRefKey),
+				},
+			},
+		}
+
+		testReconciler = &SpannerAutoscalerReconciler{
+			ctrlClient: k8sClient,
+			apiReader:  k8sClient,
+		}
+
+	})
+
+	AfterEach(func() {
+		if testSecret != nil {
+			By("Creating a Secret")
+			err := k8sClient.Create(ctx, testSecret)
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		By("Fetching credentials")
+		got, err := testReconciler.fetchCredentials(ctx, testResource)
+
+		if result.expectedErr == nil && err != nil {
+			By("Failed to fetch credentials")
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		if result.expectedErr != nil {
+			By("Comparing expected and received errors")
+			Expect(err).To(MatchError(result.expectedErr))
+		}
+
+		By("Comparing the values")
+		// using 'cmp.Diff' over 'Expect(want).To(Equal(got))' makes it
+		// easier to spot the difference when tests fail
+		Expect(cmp.Diff(result.want, got)).To(BeEmpty())
+	})
+
+	It("should fetch json correctly", func() {
+		result.want = syncer.NewServiceAccountJSONCredentials([]byte(testSecretRefVal))
+	})
+
+	It("should fetch correctly even when non-json data is provided in the secret", func() {
+		testSecret.StringData = map[string]string{testSecretRefKey: "non{json-data"}
+		result.want = syncer.NewServiceAccountJSONCredentials([]byte("non{json-data"))
+	})
```

Comment:
This is the only new test case added. Rest all are just migrations of the old tests.